File: backend-go/internal/metrics/metrics.go
package metrics

import (
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

type Collector struct {
	httpRequests   *prometheus.CounterVec
	httpDuration   *prometheus.HistogramVec
	cacheHits      *prometheus.CounterVec
	cacheMisses    *prometheus.CounterVec
	dbQueryDuration *prometheus.HistogramVec
	registry       *prometheus.Registry
	serviceName    string
}

func New(serviceName string) *Collector {
	registry := prometheus.NewRegistry()

	labels := prometheus.Labels{"service": serviceName}

	httpRequests := prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name:        "http_requests_total",
			Help:        "Total HTTP requests",
			ConstLabels: labels,
		},
		[]string{"method", "route", "status"},
	)

	httpDuration := prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:        "http_request_duration_seconds",
			Help:        "HTTP request duration",
			ConstLabels: labels,
			Buckets:     []float64{0.01, 0.05, 0.1, 0.2, 0.5, 1, 2},
		},
		[]string{"method", "route", "status"},
	)

	cacheHits := prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name:        "cache_hits_total",
			Help:        "Cache hits",
			ConstLabels: labels,
		},
		[]string{"cache"},
	)

	cacheMisses := prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name:        "cache_misses_total",
			Help:        "Cache misses",
			ConstLabels: labels,
		},
		[]string{"cache"},
	)

	dbQueryDuration := prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:        "db_query_duration_seconds",
			Help:        "Database query duration",
			ConstLabels: labels,
			Buckets:     []float64{0.001, 0.005, 0.01, 0.05, 0.1, 0.5},
		},
		[]string{"operation"},
	)

	registry.MustRegister(httpRequests, httpDuration, cacheHits, cacheMisses, dbQueryDuration)

	return &Collector{
		httpRequests:    httpRequests,
		httpDuration:    httpDuration,
		cacheHits:       cacheHits,
		cacheMisses:     cacheMisses,
		dbQueryDuration: dbQueryDuration,
		registry:        registry,
		serviceName:     serviceName,
	}
}

func (c *Collector) Handler() gin.HandlerFunc {
	h := promhttp.HandlerFor(c.registry, promhttp.HandlerOpts{})
	return func(ctx *gin.Context) {
		h.ServeHTTP(ctx.Writer, ctx.Request)
	}
}

func (c *Collector) Middleware() gin.HandlerFunc {
	return func(ctx *gin.Context) {
		start := time.Now()
		ctx.Next()
		status := ctx.Writer.Status()
		duration := time.Since(start).Seconds()
		method := ctx.Request.Method
		route := ctx.FullPath()
		if route == "" {
			route = ctx.Request.URL.Path
		}
		c.httpRequests.WithLabelValues(method, route, statusCode(status)).Inc()
		c.httpDuration.WithLabelValues(method, route, statusCode(status)).Observe(duration)
	}
}

func (c *Collector) CacheHit(cacheType string) {
	c.cacheHits.WithLabelValues(cacheType).Inc()
}

func (c *Collector) CacheMiss(cacheType string) {
	c.cacheMisses.WithLabelValues(cacheType).Inc()
}

func (c *Collector) ObserveDBQuery(operation string, duration time.Duration) {
	c.dbQueryDuration.WithLabelValues(operation).Observe(duration.Seconds())
}

func statusCode(code int) string {
	return strconv.Itoa(code)
}

File: backend-rust/src/metrics.rs
use prometheus::{
    register_histogram_vec_with_registry, register_int_counter_vec_with_registry, Encoder,
    HistogramVec, IntCounterVec, Registry, TextEncoder,
};

#[derive(Clone)]
pub struct AppMetrics {
    pub registry: std::sync::Arc<Registry>,
    pub http_requests_total: IntCounterVec,
    pub http_request_duration: HistogramVec,
    pub cache_hits: IntCounterVec,
    pub cache_misses: IntCounterVec,
    pub cache_operation_duration: HistogramVec,
    pub db_query_duration: HistogramVec,
}

impl AppMetrics {
    pub fn new() -> Self {
        let mut default_labels = std::collections::HashMap::new();
        default_labels.insert("service".to_string(), "forecap-api-rust".to_string());

        let registry = std::sync::Arc::new(
            Registry::new_custom(Some("forecap".into()), Some(default_labels)).unwrap(),
        );

        let http_requests_total = register_int_counter_vec_with_registry!(
            "http_requests_total",
            "Total HTTP requests",
            &["method", "route", "status"],
            registry.as_ref()
        )
        .unwrap();

        let http_request_duration = register_histogram_vec_with_registry!(
            "http_request_duration_seconds",
            "HTTP request duration",
            &["method", "route", "status"],
            vec![0.01, 0.05, 0.1, 0.2, 0.5, 1.0],
            registry.as_ref()
        )
        .unwrap();

        let cache_hits = register_int_counter_vec_with_registry!(
            "cache_hits_total",
            "Cache hits",
            &["cache"],
            registry.as_ref()
        )
        .unwrap();

        let cache_misses = register_int_counter_vec_with_registry!(
            "cache_misses_total",
            "Cache misses",
            &["cache"],
            registry.as_ref()
        )
        .unwrap();

        let cache_operation_duration = register_histogram_vec_with_registry!(
            "cache_operation_duration_seconds",
            "Cache operation duration",
            &["operation"],
            vec![0.001, 0.005, 0.01, 0.05, 0.1],
            registry.as_ref()
        )
        .unwrap();

        let db_query_duration = register_histogram_vec_with_registry!(
            "db_query_duration_seconds",
            "Database query duration",
            &["operation"],
            vec![0.001, 0.005, 0.01, 0.05, 0.1],
            registry.as_ref()
        )
        .unwrap();

        Self {
            registry,
            http_requests_total,
            http_request_duration,
            cache_hits,
            cache_misses,
            cache_operation_duration,
            db_query_duration,
        }
    }

    pub fn gather(&self) -> Vec<u8> {
        let metric_families = self.registry.gather();
        let mut buffer = Vec::new();
        let encoder = TextEncoder::new();
        encoder
            .encode(&metric_families, &mut buffer)
            .expect("Failed to encode metrics");
        buffer
    }
}

